Add Missing BOM Back to the `Slice/utf8BOM` Test

InsertCreativityHere opened the pull request:
This PR fixes #3387.
The Slice parser has a special grammar rule for matching and handling UTF BOMs in Slice files. And we even have a test which exists to exercise this part of the parser. Except that the Slice file we use for testing no longer has a BOM in it...

My guess is it was accidentally removed by one of our automated copyright header updates.

So, this PR adds the BOM back to the file, and adds a Python test script to the directory which:
1) Checks that the BOM is present (in case it gets deleted by accident again)
2) Runs `slice2cpp` over the file and makes sure it succeeds with no warnings/errors

----

Previously, we were just letting msbuild compile this Slice file during the test build.
Now that we call `slice2cpp` directly from the test script, having `msbuild` build it is redundant.
So I deleted all the `msbuild` files for this test directory. Which saves quite a couple lines.

InsertCreativityHere commented on cpp/test/Slice/utf8BOM/Test.ice:
```diff
@@ -1,10 +1,9 @@
-// Copyright (c) ZeroC, Inc.
+﻿// Copyright (c) ZeroC, Inc.
```
You can't see it, but there is a BOM at the start of this file.

externl commented on cpp/test/Slice/utf8BOM/test.py:
```diff
@@ -0,0 +1,42 @@
+            current.write("Checking for UTF8-BOM marker... ")
+            with open(test_file_name, 'rb') as test_file:
+                BOM = test_file.read(3)
+                if (BOM[0] != 0xEF or BOM[1] != 0xBB or BOM[2] != 0xBF):
```
Another option is something like:
```suggestion
                if BOM[:3] != b'\xEF\xBB\xBF':
```

InsertCreativityHere replied:
Fixed!

InsertCreativityHere commented:
After learning about this bug with MCPP on Ubuntu (see #3940), I updated the BOM test.
So now it has 2 BOMs, a valid one at the beginning, and an invalid one in the middle of the file.
On Windows/MacOS we expect only 1 error (for the middle one), and on linux we tolerate a 2nd error (for the beginning one).

I also added specific BOM error handling to the Slice scanner. So now it will say:
```
encountered unexpected UTF-8 BOM in input; BOMs can only appear at the beginning of files
```
instead of:
```
illegal input character: '\357'
illegal input character: '\273'
illegal input character: '\277'
```

Lastly... we were apparently testing BOMs in the `IllegalIdentifier` test too.
I removed all this because BOMs are unrelated to 'identifier's, illegal or otherwise.

InsertCreativityHere commented on cpp/src/Slice/Scanner.cpp:
Generated file. Should probably ignore.

InsertCreativityHere commented on cpp/src/Slice/Scanner.cpp:
Generated file, should ignore.

InsertCreativityHere commented on cpp/test/Slice/errorDetection/IllegalIdentifier.ice:
```diff
@@ -1,20 +1,8 @@
-
-// NOTE: Keep the initial blank line. This is necessary to avoid a problem
-// with system-provided mcpp versions that don't include ZeroC's BOM patch.
```
This comment is from 7 years ago. So I guess this has been a known issue for a while... And we even knew that having a comment on the first line triggered the re-ordering bug.

Adding this blank line at the top feels like a lame fix though.

pepone commented on cpp/test/Slice/utf8BOM/test.py:
```diff
@@ -0,0 +1,52 @@
+        test_file_name = "Test.ice"
+        slice2cpp = SliceTranslator("slice2cpp", quiet=True)
+
+        # Temporary output directory for 'slice2cpp to use.
```
Why do we need to create the output directory? I was thinking slice2cpp shouldn't generate any code here as the Slice file has errors.

InsertCreativityHere replied:
I added it just to make sure that the tests will clean up after themselves no matter what.
Although you're right that under normal circumstances, nothing will be generated.

We also create this tmp folder in the `errorDetection` tests, which is in the same situation: https://github.com/zeroc-ice/ice/blob/86dfedab1473ae2d03d65309ac78b1400229d86a/cpp/test/Slice/errorDetection/test.py#L16-L19

bernardnormier commented on cpp/src/Slice/Scanner.l:
```diff
@@ -529,8 +529,18 @@ floating_literal    (({fractional_constant}{exponent_part}?)|((\+|-)?{dec}+{expo
-<INITIAL>{bom} {}
+  /* Handles UTF-8 BOMs.
+   * The scanner has a special 'INITIAL' state, which it enters whenever it hits the start of a file.
+   * This occurrs when parsing initially begins, but also when we see the start of an '#include'd file.
```
occurs

InsertCreativityHere replied:
Good catch.
Fixed!

InsertCreativityHere commented on cpp/test/Slice/utf8BOM/test.py:
```diff
@@ -0,0 +1,52 @@
+        test_file_name = "Test.ice"
+        slice2cpp = SliceTranslator("slice2cpp", quiet=True)
+
+        # Temporary output directory for 'slice2cpp to use.
```
```suggestion
        # Temporary output directory for 'slice2cpp' to use.
```
